util/ranger/detacher.go

```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+	newConditions := make([]expression.Expression, 0, len(conditions))
+	newConditions = append(newConditions, conditions...)
+	newConditions = removeAccessConditions(newConditions, accesses)
+
+	// add Gc condition for accesses and return new condition to newAccesses
+	newAccesses, err := AddGcColumnCond(sctx, cols, accesses, columnValues)
+	if err != nil {
+		return conditions, err
+	}
+
+	// merge newAccesses and original condition execept accesses
+	newConditions = append(newConditions, newAccesses...)
+
+	return newConditions, nil
+}
+
+// NeedAddGcColumn4ShardIndex check whether to add `tidb_shard(x) = xxx`
+// @param[in] cols          the columns of shard index, such as [tidb_shard(a), a, ...]
+// @param[in] accessCond    the conditions relative to the index and arranged by the index column order.
+//                          e.g. the index is uk(tidb_shard(a), a, b) and the where clause is
+//                          `WHERE b = 1 AND a = 2 AND c = 3`, the param accessCond is {a = 2, b = 1} that is
+//                          only relative to uk's columns.
+// @param[in] columnValues  the values of index columns in param accessCond. if accessCond is {a = 2, b = 1},
+//                          columnValues is {2, 1}. if accessCond the "IN" function like `a IN (1, 2)`, columnValues
+//                          is empty.
+// @retval -  return true if it needs to addr tidb_shard() prefix, ohterwise return false
+func NeedAddGcColumn4ShardIndex(
+	cols []*expression.Column,
+	accessCond []expression.Expression,
+	columnValues []*valueInfo) bool {
+
+	// the columns of shard index shoude be more than 2, like (tidb_shard(a),a,...)
+	// check cols and columnValues in the sub call function
+	if len(accessCond) < 2 {
+		return false
+	}
+
+	if !IsValidShardIndex(cols) {
+		return false
+	}
+
+	// accessCond[0] shoudle be nil, because it has no access condition for
+	// the prefix tidb_shard() of the shard index
+	if cond := accessCond[1]; cond != nil {
+		if f, ok := cond.(*expression.ScalarFunction); ok {
+			switch f.FuncName.L {
+			case ast.EQ:
+				return NeedAddColumn4EqCond(cols, accessCond, columnValues)
+			case ast.In:
+				return NeedAddColumn4InCond(cols, accessCond, f)
+			}
+		}
+	}
+
+	return false
+}
+
+// NeedAddColumn4EqCond `tidb_shard(x) = xxx`
+// For param explanation, please refer to the function `NeedAddGcColumn4ShardIndex`.
+// It checks whether EQ condtions need to be added tidb_shard() prefix.
+// (1) columns in accessCond are all columns of the index except the first.
+// (2) every column in accessCond has a constan value
+func NeedAddColumn4EqCond(cols []*expression.Column,
+	accessCond []expression.Expression, columnValues []*valueInfo) bool {
+	valCnt := 0
+	matchedKeyFldCnt := 0
+
+	// the columns of shard index shoude be more than 2, like (tidb_shard(a),a,...)
+	if len(cols) < 2 ||
+		len(accessCond) < 2 ||
+		len(columnValues) < 2 {
+		return false
+	}
+
+	for _, cond := range accessCond[1:] {
```

Comment:
Do we need to just make sure that only the first column is eq cond?
If the index is `(shard(a), a, b)`. `a = 1 and b > 2` can also be use to construct ranges.

Reply:
We add the prefix just for point get.

Reply:
For index `shard(a), a, b`, The `a > 1` cannot be used but `a = 1 and b > 2` can.

Reply:
we support shard index just for unique secondary index and point get plan. For index `(shard(a), a, b)`, the condition  `a = 1 and b > 2` can't make a point get plan.